docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+
+  // Release allows a read lease to be released before it
+  // expires. This should be used when a watch request indicates a new
+  // version of a descriptor is available and the lease holder
+  // determines it is no longer using the old version.
+  rpc Release(ReleaseRequest) returns (ReleaseResponse) {}
+
+  // Publish publishes a table descriptor change, writing the table
+  // descriptor to the KV store on success and notifying all watches
+  // of the new version. An error is returned if the version of the
+  // table descriptor is not equal to the current version + 1. An
+  // error is returned if there are two active versions of the table
+  // descriptor.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+}
+```
+
+When a node starts up it will register a `Watch` with the lease master
+in order to receive a stream of version updates for all
+descriptors. Each node will maintain a map from `<descriptorID,
+version>` to `<TableDescriptor, expiration, localRefCount>`. When the
+node needs to use a descriptor for an operation it will look in its
+local map. If the descriptor is not there it will be read from the KV
+store and a lease will be acquired by sending an `Acquire` message to
+the lease master. The node will maintain the invariant that a new
+reference to a descriptor will only be added to the most recent
+version. When the reference count for an older version falls to 0 the
+node will send a `Release` message to the lease master.
+
+The `Watch` responses will notify the node when a new version of a
+descriptor is available which it will then read from the KV store if
+the node has a lease for the descriptor.
+
+When performing a schema change operation, the modification to the
+table descriptor is performed using the `Publish` operation which will
+fail if the version is invalid (i.e. too old) or if there are already
+two active versions of the descriptor in the cluster. On success, the
+`Publish` operation will write the new descriptor to the KV store and
+notify all watches of the new descriptor which will cause those nodes
+to stop using the old descriptor for new operations. As existing
+operations complete they will release their local reference to the old
+descriptor which will cause the reference count to drop to 0 and
+result in the lease being released. When all of the leases for the old
+descriptor have been released, the lease master will send out a
+notification to the watches that the version is `DEAD`.
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will renew recently used descriptors before their
+leases expires and ensure that when handing out a descriptor for local
+usage the lease will not expire for at least `<leaseDuration>/2`
+minutes. When a node starts it will issue a `Release` message with the
+`all` field set to true to indicate it wants to release any leases
```

Comment:
Is this safe? I'm concerned about the possibility of operations that started under the old incarnation proceeding under the authority of a lease that has been invalidated by the new incarnation. A one-phase transaction can be forwarded from its original coordinator to the raft leader where it may be delayed and retried. I think a node should persist information about leases it owns instead of issuing a blanket invalidation on recovery.